src/uchen/tboard/tboard_file.cc

```diff
@@ -2,24 +2,167 @@
+  }
+
+  // If the path is a directory, make the directory if it doesn't exist.
+  if (path.find('/') != std::string::npos) {
+    std::string_view directory = path.substr(0, path.find_last_of('/'));
+
+    if (!std::filesystem::exists(directory)) {
+      if (!std::filesystem::create_directories(directory)) {
+        return absl::InternalError("Failed to create logs directory");
+      }
+    }
+  }
+
   std::ofstream file(std::string(path), std::ios::binary);
   if (!file.is_open()) {
     return absl::InternalError("Failed to open file");
   }
   return TBoardFile(std::move(file));
 }
 
-void TBoardFile::RecordLoss(float loss) {
+void TBoardFile::RecordLoss(double loss, uint32_t step) {
   LOG(INFO) << "Recording loss: " << loss;
-  file_ << "Loss: " << loss << std::endl;
+  TBoardFile::AddScalar("loss", loss, step);
+}
+
+void TBoardFile::RecordScalar(const std::string& tag, double value,
+                              uint32_t step) {
+  TBoardFile::AddScalar(tag, value, step);
+}
+
+void TBoardFile::RecordImage(const std::string& tag, const std::string& image,
+                             uint32_t step, uint32_t width, uint32_t height,
+                             uint32_t channel, const std::string& image_name,
+                             const std::string& image_description) {
+  std::string encoded_image = EncodeImage(image);
+
+  LOG(INFO) << "Recording image: " << tag;
+
+  tensorflow::SummaryMetadata* metadata = new tensorflow::SummaryMetadata();
+  if (!image_name.empty()) {
+    metadata->set_display_name(image_name);
+  } else {
+    metadata->set_display_name(tag);
+  }
+  metadata->set_summary_description(image_description);
+
+  tensorflow::Summary::Image* image_summary = new tensorflow::Summary::Image();
+  image_summary->set_width(width);
+  image_summary->set_height(height);
+  image_summary->set_colorspace(channel);
+  image_summary->set_encoded_image_string(encoded_image);
+
+  tensorflow::Summary* summary = new tensorflow::Summary();
+  tensorflow::Summary::Value* summary_value = summary->add_value();
+  if (!summary_value) {
+    LOG(ERROR) << "[" << tag << "] Failed to add value to summary";
+    return;
+  }
+
+  summary_value->set_tag(tag);
+  summary_value->set_allocated_image(image_summary);
+  summary_value->set_allocated_metadata(metadata);
+
+  AddEvent(summary, step);
 }
 
+void TBoardFile::AddScalar(const std::string& tag, double value,
+                           uint32_t step) {
+  LOG(INFO) << "Adding scalar: " << tag << ": " << value << ", Step: " << value;
+
+  tensorflow::Summary* summary = new tensorflow::Summary();
+
+  tensorflow::Summary::Value* summary_value = summary->add_value();
+  if (!summary_value) {
+    LOG(ERROR) << "[" << tag << "] Failed to add value to summary";
+    return;
+  }
+  summary_value->set_tag(tag);
+  summary_value->set_simple_value(value);
+
+  AddEvent(summary, step);
+}
+
+void TBoardFile::AddEvent(tensorflow::Summary* summary, uint32_t step) {
```

Comment:
```suggestion
void TBoardFile::AddEvent(const tensorflow::Summary& summary, uint32_t step) {
```

Can it take `tensorflow::Summary::Value` instead? Looks like that's what we actually want.

Reply:
cannot be constant for the same reason, as `Event` will destruct summary when it's done writing it to the log file.